Customer churn ZenFile (#28)
zenml-io/zenfiles: customer-churn/README.md

So, To achieve this in a real-world scenario, we will be using [ZenML](https://zenml.io/) to build a production-ready pipeline to predict whether a customer will churn or not before they even do it.
The purpose of this repository is to demonstrate how [ZenML](https://github.com/zenml-io/zenml) empowers your business to build and deploy machine learning pipelines in a multitude of ways:

- By offering you a framework or template to develop within.
- By integrating with popular tools like Kubeflow, Seldon-core, facets, and more.
- By allowing you to build and deploy your machine learning pipelines easily using the modern MLOps Framework.

## :snake: Python Requirements [WIP]

Let's jump into the Python packages you need. Within the Python environment of your choice, run:

```bash
git clone https://github.com/zenml-io/zenfiles.git
cd zenfiles/customer-churn
pip install -r requirements.txt
```

We need to install the following integrations for this project:

```bash
zemml integration install mlflow -f
zemml integration install kubeflow -f
```

## 📙 Resources & References

> We had written a blog that explains this project in-depth: [Predicting whether the customer will churn or not before they even did it](#).

Author: Yaa Sure, I know that.

> If you'd like to watch the video that explains the project, you can watch the [video](#).

Author: Sure, I remember all of these, these to be added once videos are ready.

# :thumbsup: The Solution

We showcase two solutions to this problem:

- `Deployment using Kubeflow pipelines`: We will be using Kubeflow pipelines to build and run our ZenML pipeline on the cloud and deploy it in a production environment.
  `Continuous Deployment using Seldon Core`: We will be using Seldon Core, a production-grade open-source model serving platform, to build our continuous deployment pipeline that trains a model and then serves it with Seldon Core.

## Deploy pipelines to production using Kubeflow pipelines

To build a real-world workflow for predicting whether a customer will churn or not, you will probably develop your pipelines on your local machine initially, allowing for quick iteration and debugging. However, at a certain point, when you are finished with its design, you might want to transition to a more production-ready setting and deploy the pipeline to a more robust environment. This is where ZenML comes in.

> We will be using [Kubeflow](#) Integration of ZenML for deploying pipelines to production using Kubeflow pipelines on the cloud.

Author: Sure

- `ingest_data`: Ingest the data from the source and create a DataFrame.
- `encode_cat_cols`: Encode categorical columns.
- `handle_imbalanced_data`: Handle imbalanced data.
- `drop_cols`: Dropping irrelevant columns.
- `data_splitter`: Split the data into training and test sets.
- `model_trainer`: Train the model.
- `evaluation`: Evaluate the trained model.

#### Run the same pipeline on a local Kubeflow Pipelines deployment

With all the installation and initialization out of the way, all that's left to do is configure our ZenML stack. For this example, the stack we create consists of the following four parts:

> - The **local artifact store** stores step outputs on your hard disk.

Reviewer: It got renamed as a stack component, but it's still local as compared to a cloud orchestrator.

> - `evaluation`: This step will be used to evaluate the trained model.

Author: Any idea for it?

![cloudkubeflowstack](_assets/cloudstack.gif)

After you fulfill the prerequisites, now we need to Integrate with ZenML.

1. Install the cloud provider

```bash
zenml integration install aws
```

2. Register the stack components

```bash
zenml container-registry register cloud_registry --type=default --uri=$PATH_TO_YOUR_CONTAINER_REGISTRY
zenml orchestrator register cloud_orchestrator --type=kubeflow --custom_docker_base_image_name=YOUR_IMAGE
zenml metadata-store register kubeflow_metadata_store --type=kubeflow
zenml artifact-store register cloud_artifact_store --type=s3 --path=$PATH_TO_YOUR_BUCKET

# Register the cloud stack
zenml stack register cloud_kubeflow_stack -m kubeflow_metadata_store -a cloud_artifact_store -o cloud_orchestrator -c cloud_registry
```

3. Activate the newly created stack.

```bash
zenml stack set cloud_kubeflow_stack
```

4. Do a pipeline run

```shell
python run_kubeflow_pipeline.py
```

> 5. Check the Kubeflow UI to see if the model is deployed and running! 🚀

Author: Yeah it make sense, I will change it rn

zenml metadata-store register aws --type=kubeflow
zenml orchestrator register aws --type=kubeflow --kubernetes_context=zenml-eks --synchronous=True
zenml secrets-manager register aws -t aws
zenml stack register aws -m aws -a aws -o aws -c aws -d seldon_aws -x aws
zenml stack set aws
```

4. Do a pipeline run

```shell
python run_seldon_deployment_pipeline.py --deploy
```

You can control which pipeline to run by passing the --deploy and the --predict flag to the `run_seldon_deployment_pipeline.py` launcher. If you run the pipeline with the --deploy flag, the pipeline will train the model and deploy if the model meets the evaluation criteria and then Seldon Core will serve the model for inference. If you run the pipeline with the --predict flag, this tells the pipeline only to run the inference pipeline and not the training pipeline.

You can also set the --min-accuracy to control the evaluation criteria.

5. Check the Kubeflow UI to see if the model is deployed and running! 🚀

```bash
kubectl --namespace kubeflow port-forward svc/ml-pipeline-ui 8080:80
```

## 🕹 Demo App [WIP (To be updated)]

> **Ignore this section as of now**

Author: haha lol

To achieve this in a real-world scenario, we will be using [ZenML](https://zenml.io/) to build a production-ready pipeline that predicts whether a customer will churn or not ahead of time.
The purpose of this repository is to demonstrate how [ZenML](https://github.com/zenml-io/zenml) empowers your business to build and deploy machine learning pipelines in a multitude of ways:

- By offering you a framework or template to develop within.
- By integrating with popular tools like Kubeflow, Seldon-core, facets, and more.
- By allowing you to build and deploy your machine learning pipelines easily using the modern MLOps Framework.

## :snake: Python Requirements [WIP]

Let's jump into the Python packages you need. Within the Python environment of your choice, run:

```bash
git clone https://github.com/zenml-io/zenfiles.git
cd zenfiles/customer-churn
pip install -r requirements.txt
```

We need to install the following integrations for this project:

```bash
zemml integration install mlflow -f
zemml integration install kubeflow -f
```

## 📙 Resources & References

> We had written a blog that explains this project in-depth: [Predicting whether the customer will churn or not before they even did it](#).

Author: Sure, I will change it once it's updated

![cloudkubeflowstack](_assets/cloudstack.gif)

After you fulfill the prerequisites, now we need to Integrate with ZenML.

1. Install the cloud provider

```bash
zenml integration install aws
```

2. Register the stack components

```bash
zenml container-registry register cloud_registry --type=default --uri=$PATH_TO_YOUR_CONTAINER_REGISTRY
zenml orchestrator register cloud_orchestrator --type=kubeflow --custom_docker_base_image_name=YOUR_IMAGE
zenml metadata-store register kubeflow_metadata_store --type=kubeflow
zenml artifact-store register cloud_artifact_store --type=s3 --path=$PATH_TO_YOUR_BUCKET

# Register the cloud stack
zenml stack register cloud_kubeflow_stack -m kubeflow_metadata_store -a cloud_artifact_store -o cloud_orchestrator -c cloud_registry
```

3. Activate the newly created stack.

```bash
zenml stack set cloud_kubeflow_stack
```

4. Do a pipeline run

```shell
python run_kubeflow_pipeline.py
```

> 5. Configure Port Forwarding and Check the Kubeflow UI to see if the model is deployed and running! 🚀

Reviewer: ```suggestion
5. Configure port Forwarding and check the Kubeflow UI to see if the model is deployed and running! 🚀
```

```bash
kubectl --namespace kubeflow port-forward svc/ml-pipeline-ui 8080:80
```

> Now, you can go to the localhost:8080 to see the UI.

Reviewer: ```suggestion
Now, you can go to [the localhost URL](http://localhost:8080/#/runs) to see the UI.
```

> We can fetch the model from the kubeflow pipeline and use it in our Inference pipeline. Following diagram shows the flow of the whole pipeline:

Reviewer: ```suggestion
We can fetch the model from Kubeflow Pipelines and use it in our Inference pipeline. The following diagram shows the flow of the whole pipeline:
```

![cloudkubeflowstack](_assets/wholekubeflowstack.gif)

## Continuous model deployment with Seldon Core

> While building the real-world workflow for predicting whether a customer will churn or not, you might not want to train the model once and deploy it to production. Instead, you might want to train the model and deploy it to production when something gets triggered. This is where one of our recent integration comes in, [Seldon Core](https://github.com/zenml-io/zenml/tree/main/examples/seldon_deployment).

Reviewer: ```suggestion
While building the real-world workflow for predicting whether a customer will churn or not, you might not want to train the model once and deploy it to production. Instead, you might want to train the model and deploy it to production when something gets triggered. This is where one of our recent integrations is invaluable: [Seldon Core](https://github.com/zenml-io/zenml/tree/main/examples/seldon_deployment).
```

[Seldon Core](https://github.com/SeldonIO/seldon-core) is a production-grade open-source model serving platform. It packs a wide range of features built around deploying models to REST/GRPC microservices, including monitoring and logging, model explainers, outlier detectors, and various continuous deployment strategies such as A/B testing and canary deployments, and more.

It also comes equipped with a set of built-in model server implementations designed to work with standard formats for packaging ML models that greatly simplify serving models for real-time inference.

> In this project, we build a continuous deployment pipeline that trains a model and then serves it with Seldon Core as the industry-ready model deployment tool of choice. If you are interested in learning more about Seldon Core, you can check out the [ZenML Example](https://github.com/zenml-io/zenml/tree/main/examples/seldon_deployment). Following diagram shows the flow of the whole pipeline:

Reviewer: ```suggestion
In this project, we build a continuous deployment pipeline that trains a model and then serves it with Seldon Core as the industry-ready model deployment tool of choice. If you are interested in learning more about Seldon Core, you can check out the [ZenML example](https://github.com/zenml-io/zenml/tree/main/examples/seldon_deployment). The following diagram shows the flow of the whole pipeline:
```

![seldondeployment](_assets/seldoncorecontinousdeployment.gif)

> Now, Let's start setting up our Full AWS stack to run the pipeline using Seldon Core.

Reviewer: ```suggestion
Let's start by setting up our full AWS stack to run the pipeline using Seldon Core.
```

> 1. Install the Seldon Core Integration, a set of ZenML extensions that integrate with Seldon Core.

Reviewer: ```suggestion
1. Install the Seldon Core integration, a set of ZenML extensions that integrate with Seldon Core.
```

```bash

> zenml integration install Seldon

Reviewer: ```suggestion
zenml integration install seldon
```

```

2. Register the stack components

```bash

> aws eks --region us-east-1 update-kubeconfig --name zenml-cluster --alias zenml-eks

Reviewer: Maybe we want to give people the option to choose regions closer to them?

Author: yeah, I can add a note on changing that :)

```

To configure ECR registry access locally, run, e.g.:

```bash
aws ecr get-login-password --region us-east-1 | docker login --username AWS \

>   --password-stdin 715803424590.dkr.ecr.us-east-1.amazonaws.com

Reviewer: Is this URL meant to be hard-coded?

Author: No, it's a command

Reviewer 2: You can say it more clearly:

```
To configure ECR registry access locally, follow these [instructions](https://docs.aws.amazon.com/AmazonECR/latest/userguide/getting-started-cli.html), or simply type the following with the right  container registry address:

```bash
aws ecr get-login-password --region us-east-1 | docker login --username AWS \
  --password-stdin 715803424590.dkr.ecr.us-east-1.amazonaws.com
  ```
  ```

```

Extract the URL where the Seldon Core model server exposes its prediction API, e.g.:

```bash
export INGRESS_HOST=$(kubectl -n istio-system get service istio-ingressgateway \
  -o jsonpath='{.status.loadBalancer.ingress[0].hostname}')
```

Configuring the stack can be done like this:

```shell
zenml integration install s3 aws kubeflow Seldon

zenml artifact-store register aws --type=s3 --path=s3://mybucket
zenml model-deployer register seldon_aws --type=seldon \
  --kubernetes_context=zenml-eks --kubernetes_namespace=kubeflow \
  --base_url=http://$INGRESS_HOST \
  --secret=s3-store

> zenml container-registry register aws --type=default --uri=715803424590.dkr.ecr.us-east-1.amazonaws.com

Reviewer: Is this URI meant to be hard-coded?

Author: I suspect wdym by hard coded?

Reviewer 2: This URI is hard-coded (`715803424590.dkr.ecr.us-east-1.amazonaws.com`). Actually all of this is hard-coded. There is portion in chapter 2 of zenbytes that makes this completely variable. Check out this for reference: https://github.com/zenml-io/zenbytes/blob/main/02%20-%20Transitioning%20To%20Production.ipynb

zenml metadata-store register aws --type=kubeflow
zenml orchestrator register aws --type=kubeflow --kubernetes_context=zenml-eks --synchronous=True
zenml secrets-manager register aws -t aws
zenml stack register aws -m aws -a aws -o aws -c aws -d seldon_aws -x aws
zenml stack set aws
```

4. Do a pipeline run

```shell
python run_seldon_deployment_pipeline.py --deploy
```

> You can control which pipeline to run by passing the --deploy and the --predict flag to the `run_seldon_deployment_pipeline.py` launcher. If you run the pipeline with the --deploy flag, the pipeline will train the model and deploy if the model meets the evaluation criteria and then Seldon Core will serve the model for inference. If you run the pipeline with the --predict flag, this tells the pipeline only to run the inference pipeline and not the training pipeline.

Reviewer: ```suggestion
You can control which pipeline to run by passing the `--deploy` and the `--predict` flag to the `run_seldon_deployment_pipeline.py` launcher. If you run the pipeline with the `--deploy` flag, the pipeline will train the model and deploy if the model meets the evaluation criteria and then Seldon Core will serve the model for inference. If you run the pipeline with the `--predict` flag, this tells the pipeline only to run the inference pipeline and not the training pipeline.
```

> You can also set the --min-accuracy to control the evaluation criteria.

Reviewer: ```suggestion
You can also set the `--min-accuracy` to control the evaluation criteria.
```

> 5. Configure Port Forwarding and Check the Kubeflow UI to see if the model is deployed and running! 🚀

Reviewer: See the changes from above

Author: Sure

```bash
kubectl --namespace kubeflow port-forward svc/ml-pipeline-ui 8080:80
```

> Now, you can go to the localhost:8080 to see the UI.

Reviewer: ```suggestion
Now, you can go to the [localhost:8080](http://localhost:8080/#/runs) to see the UI.
```

## 🕹 Demo App [WIP (To be updated)]

**Ignore this section as of now**

> We also made a live demo of this project using [Streamlit](https://streamlit.io/), which you can find [here](https://share.streamlit.io/ayush714/customer-satisfaction/main). Using our trained models takes some input features for the product and predicts the customer satisfaction rate. If you want to run this streamlit app in your local system, you can run the following command:-

Reviewer: ```suggestion
We also made a live demo of this project using [Streamlit](https://streamlit.io/), which you can find [here](https://share.streamlit.io/ayush714/customer-satisfaction/main). Using our trained models takes some input features for the product and predicts the customer satisfaction rate. If you want to run this Streamlit app in your local system, you can run the following command:
```

```bash
streamlit run streamlit_app.py
```

> In the inference part of this application, we are fetching our continuous deployment pipeline from the initialized repository, and the pipeline trains the model and (re) deploys the model, and this application makes use of the latest model from the pipeline to predict the customer satisfaction score for the next order or purchase.

Reviewer: ```suggestion
In the inference part of this application, we are fetching our continuous deployment pipeline from the initialized repository. The pipeline trains the model and (re)deploys the model, and this application makes use of the latest model from the pipeline to predict the customer satisfaction score for the next order or purchase.
```

## :question: FAQ [WIP]

> To be Updated

Reviewer: Note to complete this...

Author: Yeah, I was thinking about your suggestion on making the blog post/guide on these types of question.

![cloudkubeflowstack](_assets/cloudstack.gif)

After you fulfill the prerequisites, now we need to integrate with ZenML.

1. Install the cloud provider

```bash
zenml integration install aws -y
```

2. Register the stack components

```bash
zenml container-registry register cloud_registry  --flavor=default --uri=$PATH_TO_YOUR_CONTAINER_REGISTRY
zenml orchestrator register cloud_orchestrator  --flavor=kubeflow --custom_docker_base_image_name=YOUR_IMAGE
zenml metadata-store register kubeflow_metadata_store  --flavor=kubeflow
zenml artifact-store register cloud_artifact_store  --flavor=s3 --path=$PATH_TO_YOUR_BUCKET

# Register the cloud stack
zenml stack register cloud_kubeflow_stack -m kubeflow_metadata_store -a cloud_artifact_store -o cloud_orchestrator -c cloud_registry
```

3. Activate the newly-created stack

```bash
zenml stack set cloud_kubeflow_stack
```

4. Run the pipeline

```shell
python run_kubeflow_pipeline.py
```

> 5. Configure port Forwarding and check the Kubeflow UI to see if the model is deployed and running! 🚀

Reviewer 3: ```suggestion
5. Configure port forwarding and check the Kubeflow UI to see if the model is deployed and running! 🚀
```

# Register a secret manager
zenml secrets-manager register aws_secret_manager  --flavor=aws

# Register the aws_kubeflow_stack
zenml stack register aws_kubeflow_stack -m kubeflow_metadata_store -a s3_store -o eks_orchestrator -c ecr_registry -d eks_seldon -x aws_secret_manager
```

3. Activate the newly-created stack

```bash
zenml stack set aws_kubeflow_stack
```

4. Do a pipeline run

```shell
python run_seldon_deployment_pipeline.py --secret seldon-init-container-secret --deploy
```

You can control which pipeline to run by passing the `--deploy` and the `--predict` flag to the `run_seldon_deployment_pipeline.py` launcher. If you run the pipeline with the `--deploy` flag, the pipeline will train the model and deploy if the model meets the evaluation criteria and then Seldon Core will serve the model for inference. If you run the pipeline with the `--predict` flag, this tells the pipeline only to run the inference pipeline and not the training pipeline.

You can also set the `--min-accuracy` to control the evaluation criteria.

> 5. Configure port Forwarding and check the Kubeflow UI to see if the model is deployed and running! 🚀

Reviewer 3: ```suggestion
5. Configure port forwarding and check the Kubeflow UI to see if the model is deployed and running! 🚀
```

When someone leaves a company and when that customer stops paying a business for its services or products, we call that 'churn'. We can calculate a churn rate for a company by dividing the number of customers who churned by the total number of customers and then multiplying that number by 100 to reach a percentage value. If you want to learn more about customer churn, you can read this [Wikipedia article](https://en.wikipedia.org/wiki/Churn_rate).

To achieve this in a real-world scenario, we will be using [ZenML](https://zenml.io/) to build a production-ready pipeline that predicts whether a customer will churn or not ahead of time.
The purpose of this repository is to demonstrate how [ZenML](https://github.com/zenml-io/zenml) empowers your business to build and deploy machine learning pipelines in a multitude of ways:

- By offering you a framework or template to develop within.

> - By integrating with popular and useful tools like Kubeflow, Seldon Core, `facets`, and more.

Reviewer: If you're removing facets, then you should remove it from the README.

Author: No, I am not removing facets, I am using it in data analysis pipeline, I just removed facets from requirements file which was not required.

## 📙 Resources & References

> Make sure to read the blog that introduces this project in-depth: [Predicting whether the customer will churn or not before they even did it](#).

Reviewer: ```suggestion
Make sure to read the blog that introduces this project in detail: [Predicting whether the customer will churn or not before they even did it](#).
```

If you'd like to watch the video that explains the project, you can watch the [video](#).

# :thumbsup: The Solution

We showcase two solutions to this problem:

- `Deployment using Kubeflow Pipelines`: We will be using [Kubeflow Pipelines](https://www.kubeflow.org/docs/components/pipelines/) to build and run our ZenML pipeline on the cloud and deploy it in a production environment.
- `Continuous Deployment using Seldon Core`: We will be using [Seldon Core](https://docs.seldon.io/projects/seldon-core/en/latest/index.html), a production-grade open-source model serving platform, to build our continuous deployment pipeline that trains a model and then serves it with Seldon Core.

## Deploy pipelines to production using Kubeflow Pipelines

> We will be ZenML's [Kubeflow](https://github.com/zenml-io/zenml/tree/main/examples/kubeflow) integration to deploy pipelines to production using Kubeflow Pipelines on the cloud.

Reviewer: ```suggestion
We will be using ZenML's [Kubeflow](https://github.com/zenml-io/zenml/tree/main/examples/kubeflow) integration to deploy pipelines to production using Kubeflow Pipelines on the cloud.
```

1. Install the Seldon Core integration, a set of ZenML extensions that integrate with Seldon Core.

```bash

> zenml integration install seldon -y

Reviewer: Btw, you don't always have to include the `-y` flag. All the `-y` flag does is it means you don't have to select yes or no as part of a confirmation dialogue.

Author: Sure, thanks

Configuring the stack can be done like this:

```shell

> zenml integration install s3 aws kubeflow Seldon -y

Reviewer: ```suggestion
zenml integration install s3 aws kubeflow seldon -y
```

```

3. Activate the newly-created stack

```bash
zenml stack set aws_kubeflow_stack
```

4. Do a pipeline run

```shell
python run_seldon_deployment_pipeline.py --secret seldon-init-container-secret --deploy
```

You can control which pipeline to run by passing the `--deploy` and the `--predict` flag to the `run_seldon_deployment_pipeline.py` launcher. If you run the pipeline with the `--deploy` flag, the pipeline will train the model and deploy if the model meets the evaluation criteria and then Seldon Core will serve the model for inference. If you run the pipeline with the `--predict` flag, this tells the pipeline only to run the inference pipeline and not the training pipeline.

You can also set the `--min-accuracy` to control the evaluation criteria.

5. Configure port Forwarding and check the Kubeflow UI to see if the model is deployed and running! 🚀

```bash
kubectl --namespace kubeflow port-forward svc/ml-pipeline-ui 8080:80
```

Now, you can go to the [localhost:8080](http://localhost:8080/#/runs) to see the UI.

## 🕹 Demo App

> We have made two Streamlit applications for our two different deployment solutions, one which fetches the pipeline from Kubeflow Pipelines and one which fetches the pipeline from seldon core model service.

Reviewer: ```suggestion
We built two Streamlit applications for our two different deployment solutions: one which fetches the pipeline from Kubeflow Pipelines and one which fetches the pipeline from the Seldon Core model service.
```

## 🕹 Demo App

> We built two Streamlit applications for our two different deployment solutions: one which fetches the pipeline from Kubeflow Pipelines and one which fetches the pipeline from the Seldon Core model service. You can view the Streamlit web-app [here](https://share.streamlit.io/ayush714/zenfiles/customer-churn/customer-churn/streamlit_app_kubeflow.py).

Reviewer: ```suggestion
We built two Streamlit applications for our two different deployment solutions: one which fetches the pipeline from Kubeflow Pipelines and one which fetches the pipeline from the Seldon Core model service. You can view the Streamlit web application [here](https://share.streamlit.io/ayush714/zenfiles/customer-churn/customer-churn/streamlit_app_kubeflow.py).
```
